dbstore.go

```diff
@@ -0,0 +1,199 @@
+// Copyright 2016 Mender Software AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+package main
+
+import (
+	"bytes"
+	"io"
+	"io/ioutil"
+	"os"
+	"path"
+
+	"github.com/bmatsuo/lmdb-go/lmdb"
+	"github.com/mendersoftware/log"
+	"github.com/pkg/errors"
+)
+
+const (
+	DBStoreName = "mender-store"
+)
+
+var (
+	ErrDBStoreNotInitialized = errors.New("DB store not initialized")
+)
+
+// DBStore is an opaque structure representing a database backed storage.
+// Implements `Store` interface.
+type DBStore struct {
+	env *lmdb.Env
+}
+
+type DBStoreWrite struct {
+	io.WriteCloser
+	dbs  *DBStore
+	name string
+	data bytes.Buffer
+}
+
+// NewDBStore creates an instance of Store backed by LMDB database. DBStore uses
+// a single file for DB data (named `DBStoreName`). Parameter `dirpath` is a
+// directory where the file will be stored. Returns nil if initialization
+// failed.
+func NewDBStore(dirpath string) *DBStore {
+	env, err := lmdb.NewEnv()
+	if err != nil {
+		log.Errorf("failed to create DB environment: %v", err)
+		return nil
+	}
+
+	if err := env.Open(path.Join(dirpath, DBStoreName), lmdb.NoSubdir, 0600); err != nil {
+		log.Errorf("failed to open DB environment: %v", err)
+		return nil
+	}
+
+	return &DBStore{
+		env: env,
+	}
+}
+
+func (db *DBStore) Close() error {
+	if db.env != nil {
+		if err := db.env.Close(); err != nil {
+			return errors.Wrapf(err, "failed to close DB")
+		}
+		db.env = nil
+	}
+	return nil
+}
+
+func (db *DBStore) ReadAll(name string) ([]byte, error) {
+	b, err := db.readBytes(name)
+	if err != nil {
+		return nil, err
+	}
+	return b.Bytes(), nil
+}
+
+func (db *DBStore) WriteAll(name string, data []byte) error {
+	return db.writeBytes(name, bytes.NewBuffer(data))
+}
+
+func (db *DBStore) writeBytes(name string, data *bytes.Buffer) error {
+	if db.env == nil {
+		return ErrDBStoreNotInitialized
+	}
+
+	err := db.env.Update(func(txn *lmdb.Txn) error {
```

Comment:
My point was that our `Store` API opens a new transaction for every operation. So for example, if you do `ReadAll()` followed by `WriteAll()`, this is not one transaction, but two. This can be dangerous if others are writing to the same DB (not a problem for us though), but can also be a performance bottleneck, especially for writes.

What I'm suggesting is that `Store` uses `View()` and `Update()` functions, just like LMDB does. This would allow the code that uses the DB to use transactions.

If added, I also think they should be enforced (IOW move the old `Store` read/write API into a `StoreTransaction` object, so they can only be carried out there).

So one of the functions would be something like:

```
func (self *DBStore) Update(func worker(txn *StoreTransaction)) (error)
```

What do you think?